### ISSUE TITLE
dlopen failed: library "mgba_libretro_android.so" not found

### ISSUE THREAD

**Issue by JustMeDaFaq** [Swordfish90/LibretroDroid#14]
Hi,
i just load up the project in Android Studio.
Make sure to change the hardcoded paths to existing ones in SampleActivity..

When
`retroView = GLRetroView(this, "mgba_libretro_android.so", pathtogame, "", "", LibretroDroid.SHADER_LCD)`
gets called tho, the above error gets thrown.

Does happen with any other core too.
It happens on:
```
void LibretroDroid::Core::open(const std::string& soCorePath) {
    libHandle = dlopen(soCorePath.c_str(), RTLD_LOCAL);
```

To get the detailed (in headline) error log i logged dlerror() like 
```
void LibretroDroid::Core::open(const std::string& soCorePath) {
    libHandle = dlopen(soCorePath.c_str(), RTLD_LOCAL);
    if (!libHandle) {
        LOGE("Cannot dlopen library, closing");
        LOGE("%s",dlerror());

```
Any idea? :)

**Comment by Swordfish90** [comment]
Mmmmh. This seems very strange. Could you please try renaming the library to "libmgba_libretro_android.so"? Maybe the library doesn't get extracted to the lib directory because the filename doesn't start with "lib".

**Comment by JustMeDaFaq** [comment]
Librarys got extracted correctly. It seems tho that android 7 and previous dont like to only get the librarys name.
So to work, when passing the .so on creation of the GLRetroVIew, you can just append the path to the apps native lib dir, in kotlin thats then like:
`retroView = GLRetroView(this, **getApplicationInfo().nativeLibraryDi**r+"/libdyn.so", "/storage/emulated/0/test.gba", filesDir.absolutePath, cacheDir.absolutePath, LibretroDroid.SHADER_CRT)`

**Comment by tytydraco** [comment]
Thanks for the solution, I'll be using this workaround until the LibretroDroid library is updated (unless it's a `wontfix`).

**Comment by tytydraco** [comment]
Actually, I'm getting the same error as you are when building a Release APK. Debug builds work fine. I also tried using `android:extractNativeLibs="true"` in my manifest and manually specifying the nativeLibraryDir, but it still crashes.

**Comment by tytydraco** [comment]
BIG heads up:

With release APKs, the libretro library MUST BEGIN WITH `lib`. For example, "mgba_libretro_android.so" will NOT be extracted upon launch in a release APK. However, "libmgba_libretro_android.so" WILL.

**Comment by Swordfish90** [comment]
@tytydraco Yeah, there are lots of traps in there. The library has to start with "lib" to be copied into the native library directory.

The path is passed automatically to dlopen so you can also specify the library name ("foo" is actually resolved into "libfoo.so").

But here's the kicker: the default Android behaviour changed with gradle plugin 3.6.0. Libraries are no longer installed in the native library directory, but you need to specify a flag in gradle.property AND put `extractNativeLibs=true` in Manifest.

I'm planning to write a blog post in the future since I've lost many days on the topic. :smile:
